feat: added og tags

evanessa234 opened the pull request:
This pull request makes the following changes:
- PR to resolve ushahidi/platform#4216 

Testing checklist:
- [x] Use [Ngrok](https://ngrok.com/) to test preview of the OpenGraph Meta-tags

- [x] I certify that I ran my checklist

Fixes ushahidi/platform# .

Ping @ushahidi/platform

evanessa234 commented:
> Hey @evanessa234,
> thank you for your Pull Request.
> 
> It looks like you haven't signed our **C**ontributor **L**icense **A**greement, yet.
> 
> > The purpose of a CLA is to ensure that the guardian of a project's outputs has the necessary ownership or grants of rights over all contributions to allow them to distribute under the chosen licence.
> > [Wikipedia](http://en.wikipedia.org/wiki/Contributor_License_Agreement)
> 
> You can read and sign our full Contributor License Agreement [here](https://docs.google.com/forms/d/15LyeKTOP36T5u3290o4hsmqv79v-m_s2QSmgMJXpYp8/viewform).
> 
> Once you've signed reply with `[clabot:check]` to prove it.
> 
> Appreciation of efforts,
> 
> clabot

[clabot:check]

Angamanga reviewed (changes requested):
@evanessa234 This works fine, although I am wondering if we can use the actual page title instead of "Ushahidi"? We set the title here: https://github.com/ushahidi/platform-client/blob/develop/app/index.html#L9 so maybe those variables are available to the meta-tags as well? Same for the image, it should be available in the app itself through "./img/map-new.png". I am not entirely sure if it works when it is not a `src` attribute but please give it a try :) 👍

evanessa234 commented:
> @evanessa234 This works fine, although I am wondering if we can use the actual page title instead of "Ushahidi"? We set the title here: https://github.com/ushahidi/platform-client/blob/develop/app/index.html#L9 so maybe those variables are available to the meta-tags as well? Same for the image, it should be available in the app itself through "./img/map-new.png". I am not entirely sure if it works when it is not a `src` attribute but please give it a try :)

evanessa234 commented:
@Angamanga  I am trying to fetch contents of meta tags dynamically based on the Views using this https://www.npmjs.com/package/ng-meta and https://stackoverflow.com/questions/27381918/angular-dynamic-meta-tags-in-head.

by using code as: 
```js
 .state(
        {
            url: '/views/map',
            name: 'posts.map.all',
            views: {
                'mode-context': 'modeContext'
            },
            data: {
                'meta': {
                    'title': "Tilte for /view/map",
                    'og:type': 'map',
                  'og:image': 'http://www.yourdomain.com/img/facebookimage.jpg',
                  'author': 'Evanessa'
                }
              },
            onEnter: ['$state', 'PostFilters', function ($state, PostFilters) {
                if (PostFilters.getMode() === 'savedsearch') {
                    $state.go('posts.map.savedsearch', {savedSearchId: PostFilters.getModeId()});
                } else if (PostFilters.getMode() === 'collection') {
                    $state.go('posts.map.collection', {collectionId: PostFilters.getModeId()});
                }
            }]
        }
    )
```

I tried {{pageTitle}} for og tags it didn’t work though: https://stackoverflow.com/questions/17127980/og-meta-tags-social-buttons-and-angularjs. 
I am able to load different titles and descriptions based on states, though they are hardcoded yet. I am trying to fetch them dynamically so I am currently locating view specific controllers to get scoped variables as titles, descriptions, media images (if they exist for a post). Once I locate them I will pass them in the form of variables and hence we can get dynamic meta tags.

Another challenge would be to fetch media images specific to the post if any. We cannot use them directly in the src attribute. I will see to it if something can be done using $http.

Angamanga commented:
@evanessa234 Give it a go and see if it works :) I think it's easiest to use the URL for the image you used to start with to get the title and content working at first. In that way, you won't get overwhelmed. Once the title and content are working, you can move to figure out the image 👍 Let me know if you need further guidance 🎉 .

evanessa234 commented:
> @evanessa234 Give it a go and see if it works :) I think it's easiest to use the URL for the image you used to start with to get the title and content working at first. In that way, you won't get overwhelmed. Once the title and content are working, you can move to figure out the image  Let me know if you need further guidance  .

I am able to fetch dynamic title for meta tags by using the approach given in this article https://oodavid.com/article/angularjs-meta-tags-management/ and the data is fetched from ConfigEndpoint service here https://github.com/ushahidi/platform-client/blob/develop/app/common/services/endpoints/config.js . 

But now it shows error as 
```
Error: State 'map_view' has a 'views' object. It cannot also have "view properties" at the state level.  Move the following properties into a view (in the 'views' object):  controller
``` 

below is the  change I made in https://github.com/ushahidi/platform-client/blob/develop/app/main/posts/posts-routes.js 
```
.state('map_view',
        {
            url: '/views/map',
            name: 'posts.map.all',
            views: {
                    'mode-context': 'modeContext'
            },    
            controller: ['resolvedArticle', function(resolvedArticle){
                // this.url = $location.absUrl();
                this.site = resolvedArticle;
            }],
            resolve: {
                resolvedArticle: ['$stateParams', 'ConfigEndpoint', function($stateParams, ConfigEndpoint) {    
                    let site = ConfigEndpoint.gettingMetaData().then((site) => site);
                    return site;
                  }]
            },
            onEnter: ['$state', 'PostFilters', function ($state, PostFilters) {
                if (PostFilters.getMode() === 'savedsearch') {
                    $state.go('posts.map.savedsearch', {savedSearchId: PostFilters.getModeId()});
                } else if (PostFilters.getMode() === 'collection') {
                    $state.go('posts.map.collection', {collectionId: PostFilters.getModeId()});
                }
            }]
        }
    )
```
When I put controller within view like below, the error disappears but the view is still not loaded.

```
   views: {
                    'mode-context': 'modeContext',
                     controller: ['resolvedArticle', function(resolvedArticle)
                                     {
                                                this.site = resolvedArticle;
                                    }]
            },    
           
```

I think this can be because https://github.com/ushahidi/platform-client/blob/develop/app/main/posts/views/mode-context.html is already using some other controller. I will try to modify existing controller and use it in state of ui-router. I will let you know if that works

Angamanga commented:
@evanessa234 Could you push your code even if it is not finished? That makes it easier to debug and help you out 👍

evanessa234 commented:
@Angamanga I pushed the changes, you can check now

evanessa234 commented:
okay, I missed a part of this package https://www.npmjs.com/package/ng-meta and I have figured it out now. I am trying to dynamically set meta tags using ui-router's resolve.

Angamanga commented:
@evanessa234 Perfect, let me know if you need further assistance!

evanessa234 commented:
@Angamanga , I have added meta tags for post title and content and they are working fine. For other states and routes I have added default tags in app.js

I am working on fetching map-new.png from within the app now. 
please let me know if any changes or additions are required in the pushed commits.
Thank you :)

evanessa234 commented:
@Angamanga , changes pushed. Please help review :+1:

Angamanga commented:
@evanessa234 Did you get this to work with ngrok? When I test in Facebook, I see this: 
![Screenshot 2021-04-23 at 09 06 53](https://user-images.githubusercontent.com/8624777/115832609-5f970a00-a413-11eb-8f42-5340f97e384f.png)

evanessa234 commented:
okay, let me check again

evanessa234 commented:
@Angamanga  okay sorry, I could see og meta tags being successfully fetched in head after inspecting pages....but not with facebook debugger when served with ngrok :(
I am not sure why this occured. I will google about this now. Could you point out where would I have gone wrong?

Angamanga commented:
@evanessa234 Ah, I think this actually needs to be changed in the API to work for Facebook. But the meta-tags are not bad to have anyway. I have sent this for review and discussion to the rest of the team, and we will get back to you with some more directions to unblock you. This issue turned out a bit more tricky than what it looked like in the beginning 😬 .

evanessa234 commented:
@Angamanga Thanks for looking into this and please inform me once the team decides something on this issue.

Until then can I work on any other issue?

I was looking forward to one of these issues:
1. [Remove/Uninstall Listed Unused Dependencies: #4247](https://github.com/ushahidi/platform/issues/4247)
2. [Missing Endpoint’s Response Data and Endpoints in the API documentation #4251](https://github.com/ushahidi/platform/issues/4251)

But if there are any other available issues that you would like me to work on. Can you please assign one to me?

Angamanga commented:
@evanessa234 I assigned you to https://github.com/ushahidi/platform/issues/4251, it would be fantastic if you wanted to add those!

For this PR, the progress done so far can still be used as a contribution for Outreachy, I know it's been a tricky issue and there is no obligation for PRs to be approved. If you want to keep working on it, I have discussed it with my coworkers and we came to the conclusion that some improvement is better than nothing. What you could use is adding configurable variables when the HTML is built with webpack. It's not as flexible as setting the tags on each page, but at least it's something. You could test if it is possible to use `templateParameters` in https://github.com/jantimon/html-webpack-plugin which we already use with webpack. Maybe it's possible to set the image there for all deployments and optional parameters for title and content that self-hosted deployments can make use of. Let me know if you want to test it out or need more guidance 👍

evanessa234 commented:
@Angamanga I would love to give this a try. I am currently tweaking the final application. Once I am done with it, I will resume working on this :+1:
